### ISSUE TITLE
Jax backend

### ISSUE THREAD

**Issue by MDCHAMP** [refnx/refnx#714]
Hi, really like the package. I wondered if there are any plans for a jax backend?

I noticed some comments in `reflect model.py` that seemed to indicate that this was something being considered?

I'm new to reflectrometry but would a jax backend enable autograd on the parameters of the model? Would we expect the model to be differentiable w.r.t the parameters?

**Comment by andyfaff** [comment]
Hi Max,
there is a refnx scattering kernel capable of using jax, https://github.com/refnx/refnx/blob/main/refnx/reflect/_jax_reflect.py.
The code in there is functional, although I don't think it's included in the test suite at the moment.

Unfortunately the rest of the refnx codebase is not currently capable of using jax to calculate gradients. This is because the way that parameters are set/get alters state, and jax requires statelessness. See https://jax.readthedocs.io/en/latest/jax-101/07-state.html for the kinds of things that are needed. refnx alters state to set/get because of the way it handles interparameter constraints.

Any progress on using jaxifying the codebase would probably require a complete change of architecture, and I currently don't have the skills/time to do that.

The main area where using jax to obtain automatic differentiation would lead to benefits would be using pymc samplers that use gradient information. For simple `differential_evolution` gradients aren't required.

**Comment by andyfaff** [comment]
Not really, because pretty much the entire system is stateful. You can use the reflectivity calculation kernel with jax, but not really anything else ATM.

**Comment by MDCHAMP** [comment]
Hi, I managed to get the reflectivity kernel to work in jax for our setup -- thanks for pointing me in the right direction! Gradient informaton is really speeding up the parameter identification!

In case it becomes an issue at some point, some of the functions in `_jax_reflect.py` have been depreceated.

**Comment by andyfaff** [comment]
Which of the functions are deprecated?

**Comment by andyfaff** [comment]
Also, if you have ideas for how to rearchitect refnx to be more compatible with Jax then please let me know.

**Comment by andyfaff** [comment]
@MDCHAMP please see https://github.com/refnx/refnx/pull/715, it removes the deprecated jax functions from the reflectivity kernel. You should now be able to use the most recent version of jax.
